pkg/restore.go

```diff
@@ -156,40 +170,168 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
+		}
+	}
 
-	err = session.setDatabaseCredentials(opt.kubeClient, appBinding)
-	if err != nil {
+	// update this while adding support for more backend options for backup (consul, s3, etc.)
+	if parameters.Backend != VaultStorageBackendRaft {
+		return nil, errors.New("Backend must be Raft for backup snapshots")
+	}
+
+	// clean the interim directory where the snapshot, unseal keys & root token will be stored
+	if err := clearDir(opt.interimDataDir); err != nil {
 		return nil, err
 	}
 
-	err = session.setDatabaseConnectionParameters(appBinding)
+	session := opt.newSessionWrapper(VaultCMD)
+
+	// create a new vault client to interact with vault
+	// this is needed for running commands like: vault operator raft snapshot save backup.snap or restore backup.snap
+	vaultClient, err := newVaultClient(appBinding)
 	if err != nil {
 		return nil, err
 	}
 
-	err = session.setTLSParameters(appBinding, opt.setupOptions.ScratchDir)
+	// if the vault is TLS enabled then set the env variable for vault TLS
+	if err := session.setTLSParameters(appBinding, opt.setupOptions.ScratchDir); err != nil {
+		return nil, err
+	}
+
+	// wait until the vault is ready (vault must be unsealed when ready)
+	if err := session.waitForVaultReady(vaultClient, opt.waitTimeout); err != nil {
+		return nil, err
+	}
+
+	// set the vault token that has the necessary permission to save or restore snapshot
+	if err := session.setVaultToken(opt.kubeClient, appBinding, parameters.BackupTokenSecretRef); err != nil {
+		return nil, err
+	}
+
+	// set the vault connection parameters, essentially the vault leader node address
+	if err := session.setVaultConnectionParameters(vaultClient, appBinding); err != nil {
+		return nil, err
+	}
+
+	klog.Infof("Trying to restore snapshot for VaultServer %s/%s\n", appBinding.Namespace, appBinding.Name)
+
+	opt.restoreOptions.RestorePaths = []string{opt.interimDataDir}
+
+	resticWrapper, err := restic.NewResticWrapper(opt.setupOptions)
 	if err != nil {
 		return nil, err
 	}
 
-	err = session.waitForDBReady(opt.waitTimeout)
+	restoreOutput, err := resticWrapper.RunRestore(opt.restoreOptions, targetRef)
 	if err != nil {
 		return nil, err
 	}
 
+	// restore the vault snapshot from the interim directory running command like: vault operator raft snapshot restore backup.snap
+	// for different vault server using -force flag like: vault operator raft snapshot restore -force backup.snap
+	if err := opt.restoreVaultSnapshot(session); err != nil {
+		return nil, err
+	}
+
+	// if -force is true, we're potentially dealing with restoring snapshot on a different vault server
+	// we must replace the current vault server's unseal keys & root token with the older ones that we saved during snapshot
+	if opt.force {
+		klog.Infoln("Potentially different VaultServer. Trying to migrate old unseal keys & root token.")
+		if err := opt.setVaultTokenKeys(appBinding, parameters); err != nil {
```

Comment:
You can rename `setVaultTokenKeys()` to `migrateVaultTokenKeys()` `overwriteVaultTokensKeysWithOldTokens()`